### ISSUE TITLE
Adding clang-format config and enforcing it

### ISSUE THREAD

**Issue by evelikov** [weechat/weechat#2188]
### Question

While browsing through the codebase, I noticed there are plenty of variations from what's in the docs.

So I'm wondering what's the maintainer's view about adding a clang-format config in tree (top-level .clang-format file) alongside a CI check to ensure any new commits are consistent with it, via `git-clang-format`.

I am suggesting clang-format over alternatives, since it is far more versatile and flexible. Although I am not that strongly attached to it ;-)

Fixing the style in the whole codebase (some ~250K loc) is likely to cause a lot of noise, so my gut suggestion is to avoid it if possible.

As an example I can grab the one off kernel.org and hammer it, until resembles something we're happy with.

Is that something of interest?

**Comment by flashcode** [comment]
Hi,

Yes this can be interesting to have this.

For sure the entire codebase will not be reformatted with this tool, otherwise it would break history everywhere.

And as a first step we could add the clang format file without enforcing anything.

**Comment by flashcode** [comment]
Yes the first step is to have the file to assist checking style locally before pushing.

Next step will be to enforce this, but being too strict can be discussed/challenged anyway.

**Comment by trygveaa** [comment]
I would love this! Having to ensure you're writing the correct code style manually is very tedious.

I would also definitely recommend reformatting the entire repo in one commit. That makes future changes less noisy and easier to read since you won't have style changes together with the actual changes. To avoid breaking git blame you can put the commit sha of the reformatting commit in a [ignoreRevsFile](https://git-scm.com/docs/git-config#Documentation/git-config.txt-blameignoreRevsFile).

> For sure the entire codebase will not be reformatted with this tool, otherwise it would break history everywhere.

What do you mean by break history everywhere? Are you thinking of git blame, or something else?

**Comment by flashcode** [comment]
Yes I was thinking to git blame, it makes harded to browse history of a file or lines.

Once we'll have the config file we can check the impacts of reformatting whole code, but I'm not sure this is an issue to have some code not following the style, it's more important for me to push new code with the right style.

Yes the ignoreRevsFile could be used, but if we can do without, better for me.

**Comment by trygveaa** [comment]
> I'm not sure this is an issue to have some code not following the style

It depends on how large the formatting changes are, but I really think keeping formatting changes to separate commits makes it a lot easier to read diffs.

How are you planning to check that the code follows the new code style if not all code is formatted correctly? I guess you'd have to pass only the files you've changed to the checker, or have some sort of comment in each file indicating that the file should be checked. Sounds cumbersome to me.
